src/devices/Adxl357/Adxl357.cs

```diff
@@ -0,0 +1,211 @@
+
+        /// <summary>
+        /// Gets the current temperature
+        /// </summary>
+        public Temperature Temperature => Temperature.FromDegreesCelsius(GetTemperature());
+
+        /// <summary>
+        /// Calibrates the accelerometer.
+        /// </summary>
+        /// <remarks>
+        /// Make sure that the sensor is placed horizontally when executing this method.
+        ///</remarks>
+        public void CalibrateAccelerationSensor()
+        {
+            _caliBuffer["x"] = new float[CalibrationBufferLength];
+            _caliBuffer["y"] = new float[CalibrationBufferLength];
+            _caliBuffer["z"] = new float[CalibrationBufferLength];
+
+            for (int i = 0; i < CalibrationBufferLength; i++)
+            {
+                var acc = GetRawAccelerometer();
+                _caliBuffer["x"][i] = acc.X;
+                _caliBuffer["y"][i] = acc.Y;
+                _caliBuffer["z"][i] = acc.Z;
+
+                Thread.Sleep(CalibrationInterval);
+            }
+
+            foreach (var buffer in _caliBuffer)
+            {
+                _caliData[buffer.Key] = buffer.Value.Average();
+
+                if (buffer.Key == "z")
+                {
+                    _caliData[buffer.Key] -= 10;
+                }
+            }
+
+            var x = (((_caliData["z"] - _caliData["x"]) + (_caliData["z"] - _caliData["y"])) / 2);
+            _factory = 1.0F / x;
+        }
+
+        private void SetAccelerometerRange(AccelerometerRange accelerometerRange)
+        {
+            var currentValue = ReadByte(Register.SET_RANGE_REG_ADDR);
+            var newValue = currentValue | (byte)accelerometerRange;
+            WriteRegister(Register.SET_RANGE_REG_ADDR, (byte)newValue);
+        }
+
+        private double GetTemperature()
+        {
+            Span<byte> data = stackalloc byte[2] { 0, 0 };
+
+            ReadBytes(Register.TEMPERATURE_REG_ADDR, data);
+
+            double value = ((uint)data[0] << 8) | data[1];
+
+            return 25 + (value - 1852) / -9.05;
+        }
+
+        private void Reset()
+        {
+            WriteRegister(Register.RESET_REG_ADDR, 0x52);
+            Thread.Sleep(100);
+        }
+
+        private void PowerOn()
+        {
+            WriteRegister(Register.POWER_CTR_REG_ADDR, 0x00);
+            Thread.Sleep(100);
+        }
+
+        private Vector3 GetRawAccelerometer()
+        {
+            Span<byte> data = stackalloc byte[9] { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+
+            var ace = new Vector3();
+
+            if (CheckDataReady())
+            {
+                ReadBytes(Register.FIFO_DATA_REG_ADDR, data);
+
+                double x = ((uint)data[0] << 12) | ((uint)data[1] << 4) | ((uint)data[2] >> 4);
+                double y = ((uint)data[3] << 12) | ((uint)data[4] << 4) | ((uint)data[5] >> 4);
+                double z = ((uint)data[6] << 12) | ((uint)data[7] << 4) | ((uint)data[8] >> 4);
+
+                if (((uint)x & 0x80000) == 0x80000)
```

Comment:
I feel like maybe you should use an intermediate value which is uint, make your test and then create the double one
Somethins lile:
```csharp
uint temp = ((uint)data[0] << 12) | ((uint)data[1] << 4) | ((uint)data[2] >> 4);
temp = (temp & 0x80000) == 0x80000) ? temp & 0x7ffff) - 0x80000 : temp;
ace.X = temp * _factory;
temp = etc, et, etc
ace.Y= temp * _factory;
etc
```

Reply:
I refactored the whole function. Please have a look and tell me your opinion.